### PR TITLE
Source

### PR DESCRIPTION
**Pull request by tomasz-biernacki**
Initial functionality for `source`. Needs some adjustments (probably is related to other issues).
Now these two don't work: `(source when)` and 

```
(require '[clojure.set :as s])
(source s/union)
```

### CODE REVIEW

**Comment by arichiardi** [comment]
They are failing because:

this is a `:pre` check in `validated-echo-cb`
`(or (and (find res :value) (string? (get res :value))) (and (find res :error) (instance? js/Error (get res :error))`
that assert that our answer either has value or :error
and their value must be a string or a `js/error`
BUT
`read-file....` passes back a pure nil that with `no-pr-str...` opts is not converted to string, so you don't have `:value` with a string inside.

File: src/cljs/replumb/repl.cljs [hunk 1]
```diff
@@ -499,6 +500,40 @@
...
+                                         cb)))))
+
+(defn process-source
+  [opts cb data env sym]
+  (let [var (get-var opts env sym)
+        cb-partial (partial call-back! (merge opts {:no-pr-str-on-value true}) cb data)
+        call-back (fn [result]
+                    (if (nil? result)   ; if nil? means that no file was found
+                      (cb-partial (common/wrap-success "nil"))
+                      (cb-partial result)))]
```
Comment (arichiardi) [inline src/cljs/replumb/repl.cljs]: Is `result` here a string? If yes, try to `wrap-success` here too...

File: src/cljs/replumb/repl.cljs [hunk 2]
```diff
@@ -499,6 +500,40 @@
...
+                                         meta 
+                                         :source
+                                         common/wrap-success
+                                         cb)))))
+
+(defn process-source
+  [opts cb data env sym]
+  (let [var (get-var opts env sym)
+        cb-partial (partial call-back! (merge opts {:no-pr-str-on-value true}) cb data)
+        call-back (fn [result]
```
Comment (arichiardi) [inline src/cljs/replumb/repl.cljs]: Maybe `src-cb` like the name in the README? Or `file-cb` is clear too...

Comment (arichiardi) [inline src/cljs/replumb/repl.cljs]: No sorry which callback is this? It is not `src-cb` forget about above

File: src/cljs/replumb/repl.cljs (hunk 1 above)
Comment (tomasz-biernacki) [inline src/cljs/replumb/repl.cljs]: The else branch should already receive a wrapped result, see lines above (`->`..)
But yes, maybe better wrap in one place.

**Comment by arichiardi** [comment]
See also [@mfikes](http://blog.fikesfarm.com/posts/2015-12-12-clojurescript-macro-functions.html) post on why `(source when)` is not working. Maybe we can add a failing test to show this behavior as well.
And I would open an issue for `s/union`...

**Comment by tomasz-biernacki** [comment]
Going to check because on `planck` it's working. If I don't find anything soon I will open an issue.

**Pull request merged**